src/mds/SessionMap.h

```diff
@@ -255,28 +262,61 @@ class Session : public RefCountedObject {
 
 class MDS;
 
-class SessionMap {
+/**
+ * Encapsulate the serialized state associated with SessionMap.  Allows
+ * encode/decode outside of live MDS instance.
+ */
+class SessionMapStore {
+protected:
+    mds_rank_t rank;
 public:
-  MDS *mds;
-private:
   ceph::unordered_map<entity_name_t, Session*> session_map;
+  version_t version;
+
+  virtual void encode(bufferlist& bl) const;
+  virtual void decode(bufferlist::iterator& blp);
+  void dump(Formatter *f) const;
+
+  Session* get_or_add_session(const entity_inst_t& i) {
+    Session *s;
+    if (session_map.count(i.name)) {
+      s = session_map[i.name];
+    } else {
+      s = session_map[i.name] = new Session;
+      s->info.inst = i;
+      s->last_cap_renew = ceph_clock_now(g_ceph_context);
+    }
+
+    return s;
+  }
+
+  static void generate_test_instances(list<SessionMapStore*>& ls);
+
+  void reset_state()
+  {
+    session_map.clear();
+  }
+
+  SessionMapStore() : rank(MDS_RANK_NONE), version(0) {}
```

Comment:
Actually scratch that, having a mandatory init() makes it a pain to handle SessionMap along with the other table types generically from the tabletool.  Coming back to the original reason the rank attribute existed at all (log messages use the rank), I've instead added a virtual get_rank() method that SessionMap implements via mds->whoami, and a bare SessionMapStore will always say NONE for.  Logically it makes sense as this is the *active* MDS that's holding the map, rather than being the MDS rank of the map itself: when we are an MDS we have the rank, when we are running as tabletool we are a rankless thing.

Still want to come back and address the first two paras in the previous comment in future work though.